rfcs/0092-plan-dynamism.md

```diff
@@ -0,0 +1,258 @@
+> but by removing the weaknesses and restrictions that make additional features appear necessary.
+
+We already have a dynamic scheduler that doesn't need to know all the goals up front.
+We also already rewrite derivations based on previous builds for CA-derivations.
+All the underlying mechanisms are thus there, and the patch implementing this in a sense wrote itself.
+
+Now, there is a good argument that maybe the Nix derivation language today has other implementation strategies where this *wouldn't* be so natural and easy.
+This is like saying "we can add this axiom for free in our current model, but not in all possible models of our current axioms".
+Well, if such a concrete other strategy ever arises, it is very easy to statically prohibit the new features this RFC proposes.
+Until then, down with the artificial restrictions!
+
+# Detailed design
+[design]: #detailed-design
+
+We can break this down nicely into steps.
+
+*This is implemented in https://github.com/NixOS/nix/pull/4628.*
+
+1. Derivation outputs can be valid derivations.
+   \[If one tries to output a drv file today, they will find Nix doesn't accept the output as such because these small paper cuts.
+   This list item and its children should be thought of as "lifting artificial restrictions".\]
+
+   1. Allow derivation outputs to be content-addressed in the same manner as drv files.
+      (`outputHashMode = "text";`, see [Advanced Attributes](https://nixos.org/manual/nix/unstable/expressions/advanced-attributes.html)).
+
+   2. Lift the (perhaps not yet documented) restriction barring derivations output paths from ending in `.drv`, but only for derivation outputs that are so content-addressed.
+      \[There are probably other ways to make store paths that end in `.drv` that aren't valid derivations, so we could make the simpler change of lifting this restriction entirely without breaking invariants. But I'm fine keeping it for the wrong sorts of derivations as a useful guard rail.\]
+
+2. Extend the CLI to take advantage of such derivations:
+
+   We hopefully will soon allow CLI "installable" args in the form
+   ```
+   single-installable ::= <path> ! <output-name>
+   ```
+   where the first path is a derivation, and the second is the output we want to build.
+
+   We should generalize the grammar like so:
+   ```
+   single-installable ::= <single-installable> ! <output-name>
+                       |  <path>
+
+   multi-installable  ::= <single-installable>
+                       |  <single-installable> ! *
+   ```
+
+   Plain paths just mean that path itself is the goal, while `!` indexing indicates one more outputs of the derivation to the left of the `!` is the goal.
+
+   > For example,
+   > ```
+   > nix build /nix/store/…foo.drv
+   > ```
+   > would just obtain `/nix/store/…foo.drv` and not build it, while
+   > ```
+   > nix build /nix/store/…foo.drv!*
+   > ```
+   > would obtain (by building or substituting) all its outputs.
```

Comment:
What if you want to build *just* the first layer?

I think there's another potential solution here: `nix build` builds only the derivation specified (like right now), but it is extended such that it can build derivations at floating paths.

E.g. if we build the following code:
```nix
rec {
  type = "derivation";
  drvPath = builtins.outputOf something "out";
  outPath = builtins.outputOf drvPath "out";
}
```

Then `nix build` could perhaps automatically build the prerequisite for the floating path.

Reply:
See https://github.com/NixOS/rfcs/pull/92#discussion_r671729812.

The suggestion [below that](https://github.com/NixOS/rfcs/pull/92#discussion_r671741003) would return something akin to what @L-as describes succinctly in their example.
@Ericson2314 I think I could write out that `mkPackage` function if you think that's helpful at this stage.

Reply:
There are also advanced situations where downstream derivations want to be sure they depend on "intermediate" derivations.

For example, if one is using GHC to find the dependencies between between modules of a package, we do need to know what packages are provided by upstream libraries, but we don't need to have actually *built* those upstream libriaries. I am would want to run the dep solver on the current library, depending on just the "module manifest" to allow it not to fail on upstream library `import`s, and then create derivations that depend on the upstream derivations, for maximum incrementality.

Long story short, supporting that example means situations like:
- `bar` depends on `foo!out`: drv generater `bar` depends on drv created by `foo`
- `bar!out` depends on `foo!out!out`: drv generated by var depends on result of building the drv generated by `foo`